Extend runbook for MimirIngesterReachingTenantsLimit (#4744)
grafana/mimir: docs/sources/mimir/operators-guide/mimir-runbooks/_index.md

This alert fires when the `max_tenants` per ingester instance limit is enabled and the actual number of tenants in an ingester is reaching the limit. Once the limit is reached, writes to the ingester will fail (5xx) for new tenants, while they will continue to succeed for previously existing ones.

> The per-tenant memory utilisation in ingesters is made up mostly by allocations for TSDB stripes and chunk writer buffers. The size of these allocations is controlled by `-blocks-storage.tsdb.stripe-size` (default 16KiB) and `-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes` (default 4MiB), respectively.

Reviewer: newline before "in case of emergency" section:

```suggestion
The per-tenant memory utilisation in ingesters is made up mostly by allocations for TSDB stripes and chunk writer buffers. The size of these allocations is controlled by `-blocks-storage.tsdb.stripe-size` (default 16KiB) and `-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes` (default 4MiB), respectively.

```

This alert fires when the `max_tenants` per ingester instance limit is enabled and the actual number of tenants in an ingester is reaching the limit. Once the limit is reached, writes to the ingester will fail (5xx) for new tenants, while they will continue to succeed for previously existing ones.

> The per-tenant memory utilisation in ingesters is made up mostly by allocations for TSDB stripes and chunk writer buffers. The size of these allocations is controlled by `-blocks-storage.tsdb.stripe-size` (default 16KiB) and `-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes` (default 4MiB), respectively.

Reviewer 2: I don't think we can generalize about ingester memory utilization mostly being made up of TSDB stripes and chunk writer buffers, we just spotted it with many small tenants. How about the following?
```suggestion
The per-tenant memory utilisation in ingesters includes the overhead of allocations for TSDB stripes and chunk writer buffers. If the tenant number is high, this may contribute significantly to the total ingester memory utilization. The size of these allocations is controlled by `-blocks-storage.tsdb.stripe-size` (default 16KiB) and `-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes` (default 4MiB), respectively.
```

1. Ensure shuffle-sharding is enabled in the Mimir cluster
1. Assuming shuffle-sharding is enabled, scaling up ingesters will lower the number of tenants per ingester. However, the effect of this change will be visible only after `-blocks-storage.tsdb.close-idle-tsdb-timeout` period so you may have to temporarily increase the limit

> 1. For small tenants, default allocations for TSDB stripes and chunk writer buffers may be oversized. If a cell contains only small tenants, these allocation sizes can be reduced to allow raising the `max_tenants` limit without causing greatly increased per-tenant memory consumption.

Reviewer 2: I think the following phrasing makes clearer the alternative strategy of increasing the `max tenants` limit, and how one might avoid excessive memory resulting usage:
```suggestion
1. If the cell's number of series per ingester is below our target, we may want to increase the `max tenants` limit instead of adding more ingesters (which would make the cell over-provisioned in effect). If memory usage is high (due to the number of tenants), we might want to bring it down by reducing TSDB stripe size (`-blocks-storage.tsdb.stripe-size`) and chunk writer buffers (`-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes`). The latter will have a negative performance impacts for large tenants, but for smaller tenants it should be OK. See [example PR](https://github.com/grafana/deployment_tools/pull/64936).
```

1. Ensure shuffle-sharding is enabled in the Mimir cluster
1. Assuming shuffle-sharding is enabled, scaling up ingesters will lower the number of tenants per ingester. However, the effect of this change will be visible only after `-blocks-storage.tsdb.close-idle-tsdb-timeout` period so you may have to temporarily increase the limit

> 1. If the cell's number of series per ingester is below our target, we may want to increase the `max tenants` limit instead of adding more ingesters (which would make the cell over-provisioned in effect). If memory usage is high (due to the number of tenants), we might want to bring it down by reducing TSDB stripe size (`-blocks-storage.tsdb.stripe-size`) and chunk writer buffers (`-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes`). The latter will have a negative performance impact for large tenants, but for smaller tenants it should be OK.

Reviewer 3: What we did today at Grafana Labs was a sort of snowflake change, but I don't feel comfortable adding it to the runbook, because there are many nuances changing this. Also changing both settings negatively affect large tenants. I would just remove this entire point from the list.

> 1. If the cell's number of series per ingester is below our target, we may want to increase the `max tenants` limit instead of adding more ingesters (which would make the cell over-provisioned in effect). If memory usage is high (due to the number of tenants), we might want to bring it down by reducing TSDB stripe size (`-blocks-storage.tsdb.stripe-size`) and chunk writer buffers (`-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes`). The latter will have a negative performance impacts for large tenants, but for smaller tenants it should be OK.

Reviewer 2: I spotted a typo
```suggestion
1. If the cell's number of series per ingester is below our target, we may want to increase the `max tenants` limit instead of adding more ingesters (which would make the cell over-provisioned in effect). If memory usage is high (due to the number of tenants), we might want to bring it down by reducing TSDB stripe size (`-blocks-storage.tsdb.stripe-size`) and chunk writer buffers (`-blocks-storage.tsdb.head-chunks-write-buffer-size-bytes`). The latter will have a negative performance impact for large tenants, but for smaller tenants it should be OK.
```
